### PR TITLE
Add TR-502MSV Protocol

### LINKED ISSUE #3748: tr-502msv transmitting not working
### The problem

Hi everyone, I have this RF remote from LIDL: tr-502msv that controls some RF sockets:
remote
I can receive his messages:
lidl
But I can't reproduce them to turn on and off the plugs.
This is my configuration:
config
If I activate the switch nothing happens. If I change protocol to another one, like 5 or 6, I can receive something but without the last "1"
Is it a bug or I'm just interpreting in a wrong way the remote messages?
Thanks in advance

### Which version of ESPHome has the issue?

2022.10.0

### What type of installation are you using?

pip

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP32

### Board

esp32doit-devkit-v1

### Component causing the issue

rf receiver / rf transmitter

### Example YAML snippet

```yaml
remote_receiver:
  - id: RF_RX
    pin: 
      number: GPIO5
      inverted: True
    dump: 
      - rc_switch
    tolerance: 35%
    filter: 200us
    idle: 4ms
    buffer_size: 3kb

remote_transmitter:
  - id: RF_TX
    pin: GPIO18
    carrier_duty_percent: 100%

switch:
  - platform: template
    name: RF Power Button
    turn_on_action:
      - remote_transmitter.transmit_rc_switch_raw:
          transmitter_id: RF_TX
          protocol: 2
          code: '10110000000000010001'
```

### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Mat931**
# What does this implement/fix?

Add support for the TR-502MSV RF protocol

[Transmitter documentation](https://deploy-preview-2903--esphome.netlify.app/components/remote_transmitter.html#remote-transmitter-transmit-tr-502msv-action)
[Receiver documentation](https://deploy-preview-2903--esphome.netlify.app/components/remote_receiver.html?highlight=tr_502msv)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/3748

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2903

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
external_components:
  - source: github://pr#4791
    components: [remote_base]
    refresh: always

remote_receiver:
  filter: 15us
  pin:
    number: 26
    mode:
      input: True
      pullup: True
    inverted: False
  dump: [tr_502msv, raw]

remote_transmitter:
  pin: 25
  carrier_duty_percent: 100

binary_sensor:
  - platform: gpio
    pin:
      number: 0
      inverted: true
    id: test
    filters:
      - delayed_on_off: 100ms
    on_press:
      - remote_transmitter.transmit_tr_502msv:
          group: 0x123
          device: all
          command: turn_on
  - platform: remote_receiver
    name: "Remote 1 On"
    internal: yes
    tr_502msv:
      group: 0xABC
      device: 1
      command: turn_on
    filters:
      delayed_off: 1s
    #on_press:
    #  - switch.template.publish:
    #      id: Remote1
    #      state: ON
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/remote_base/tr_502msv_protocol.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    } else {
+      dst->space(BIT_ZERO_SPACE_US);
+      dst->mark(BIT_ZERO_MARK_US);
+    }
+  }
+}
+
+optional<TR502MSVData> TR502MSVProtocol::decode(RemoteReceiveData src) {
+  TR502MSVData out;
+  out.set_raw(0);
```
Comment (jesserockz) [inline esphome/components/remote_base/tr_502msv_protocol.cpp]: I don't see the need for this when `set_raw` is called below, or `out` is not even returned.
```suggestion
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — DISMISSED** [review]
I have made some changes to the struct to use a packed union to get and set the raw value without having to bitshift everytime. Please let me know if that still works for you.

**Comment by Mat931** [comment]
Thanks for taking a look!
Unfortunately I'm getting some compile errors on the ESP8266 with your changes. My config is different from the example above.
```
In file included from src/esphome.h:63,
                 from src/main.cpp:3:
src/esphome/components/remote_base/tr_502msv_protocol.h:26:12: note: offset of packed bit-field 'esphome::remote_base::TR502MSVData::<unnamed union>::<unnamed struct>::command' has changed in GCC 4.4
   26 |     struct {
      |            ^
src/main.cpp: In function 'void setup()':
src/main.cpp:756:3: error: too many initializers for 'esphome::remote_base::TR502MSVData'
  756 |   });
      |   ^
src/main.cpp:788:3: error: too many initializers for 'esphome::remote_base::TR502MSVData'
  788 |   });
      |   ^
src/main.cpp:820:3: error: too many initializers for 'esphome::remote_base::TR502MSVData'
  820 |   });
      |   ^
src/main.cpp:852:3: error: too many initializers for 'esphome::remote_base::TR502MSVData'
  852 |   });
      |   ^
src/main.cpp:892:3: error: too many initializers for 'esphome::remote_base::TR502MSVData'
  892 |   });
      |   ^
src/main.cpp:932:3: error: too many initializers for 'esphome::remote_base::TR502MSVData'
  932 |   });
      |   ^
```

In Line 752-756 of main.cpp:
```
  remote_base_tr502msvbinarysensor->set_data(remote_base::TR502MSVData{
      .group = 0xABC,
      .device = remote_base::DEVICE_1,
      .command = remote_base::COMMAND_ON,
  });
```

**Comment by Mat931** [comment]
I updated the yaml example so you can reproduce the error

**Comment by jesserockz** [comment]
Ok, ill take another look at this locally when I get some time.

**Comment by Mat931** [comment]
Closed because it wasn't merged after a year and I don't really need it anymore.

**Pull request closed without merging**